crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -206,13 +206,51 @@ impl<'db> SemanticIndexBuilder<'db> {
     }
 
     fn pop_scope(&mut self) -> FileScopeId {
-        let ScopeInfo { file_scope_id, .. } =
-            self.scope_stack.pop().expect("Root scope to be present");
-        let children_end = self.scopes.next_index();
-        let scope = &mut self.scopes[file_scope_id];
-        scope.descendents = scope.descendents.start..children_end;
         self.try_node_context_stack_manager.exit_scope();
-        file_scope_id
+
+        let ScopeInfo {
+            file_scope_id: popped_scope_id,
+            ..
+        } = self
+            .scope_stack
+            .pop()
+            .expect("Root scope should be present");
+
+        let mut scopes = std::mem::take(&mut self.scopes);
+        let children_end = scopes.next_index();
+        let popped_scope = &mut scopes[popped_scope_id];
+        popped_scope.extend_descendents(children_end);
+
+        // If the scope that we just popped off is an eager scope, we need to "lock" our view of
+        // which bindings and/or declarations reach each of the uses in the scope.
+        if let Some(eager_nested_scope) = popped_scope.node().as_eager_nested_scope() {
+            for nested_symbol in self.symbol_tables[popped_scope_id].symbols() {
+                // Loop through each enclosing scope, looking for the innermost one that binds or
+                // declares the symbol.
+                for enclosing_scope_info in self.scope_stack.iter().rev() {
+                    let enclosing_scope_id = enclosing_scope_info.file_scope_id;
```

Comment:
I see why we need this logic here, even though it's a partial duplication of the logic in `TypeInferenceBuilder::infer_name_load`, which is a bit unfortunate. But it will be important that they stay in sync. One thing I don't see here is the special case for class scopes, which are never visible to nested function-like scopes. So even if there is a binding for `x` in a class scope, a reference to `x` in a nested comprehension should not snapshot `x` in the class scope, but rather in some ancestor scope of the class scope. Maybe we need some tests around this?

Reply:
As part of reworking how we store eager bindings, this logic now lives entirely here in the semantic index builder.  It handles stopping as soon as we encounter a lazy enclosing scope.  Type inference uses the presence of an eager bindings snapshot to indicate that the name resolution should be done eagerly in that enclosing scope.

> One thing I don't see here is the special case for class scopes, which are never visible to nested function-like scopes. So even if there is a binding for `x` in a class scope, a reference to `x` in a nested comprehension should not snapshot `x` in the class scope, but rather in some ancestor scope of the class scope. Maybe we need some tests around this?

Added a test for this